rfcs/2023-05-03-data-volume-metrics.md

```diff
@@ -0,0 +1,146 @@
+and `component_sent_event_bytes_total` metrics are the estimated JSON size of the event.
+
+#### `component_received_event_bytes_total`
+
+This metric is emitted by the framework [here][source_sender], so it looks like the only change needed is
+to add the service tag.
+
+#### `component_sent_event_bytes_total`
+
+For stream based sinks this will typically be the byte value returned by `DriverResponse::events_sent`.
+
+Despite being in the [Component Spec][component_spec], not all sinks currently conform to this.
+
+As an example, from a cursory glance over a couple of sinks:
+
+The Amqp sink currently emits this value as the length of the binary data that is sent. By the time the data has
+reached the code where the `component_sent_event_bytes_total` event is emitted, that event has been encoded
+and the actual estimated JSON size has been lost. The sink will need to be updated so that when the event is
+encoded, the encoded event together with the pre-encoded JSON bytesize will be sent to the service where the
+event is emitted.
+
+The Kafka sink also currently sends the binary size, but it looks like the estimated JSON bytesize is easily
+accessible at the point of emitting, so would not need too much of a change.
+
+To ensure that the correct metric is sent in a type-safe manner, we will wrap the estimated JSON size in a
+newtype:
+
+```rust
+pub struct JsonSize(usize);
+```
+
+The `EventsSent` metric will only accept this type.
+
+## Rationale
+
+The ability to visualize data flowing through Vector will allow users to ascertain
+the effectiveness of the current use of Vector. This will enable users to
+optimise their configurations to make the best use of Vector's features.
+
+## Drawbacks
+
+The additional tags being added to the metrics will increase the cardinality of
+those metrics if they are enabled.
+
+## Prior Art
+
+
+## Alternatives
+
+We could use an alternative metric instead of estimated JSON size.
+
+- *Network bytes* This provides a more accurate picture of the actual data being received
+  and sent by Vector, but will regularly produce different sizes for an incoming event
+  to an outgoing event.
+- *In memory size* The size of the event as held in memory. This may be more accurate in
+  determining the amount of memory Vector will be utilizing at any time, will often be
+  less accurate compared to the data being sent and received which is often JSON.
+
+## Outstanding Questions
+
+- Should the default be to not emit tags? Could the default be different for OP users and standard Vector
+  users?
+- What should the source be tagged as for events that go through a transform that could potentially
+  combine one event from multiple sources - eg. `reduce`?
+- What should the source be for events that come from no source (emitted by the `lua` transform)?
+- Not all sources will have a `service`. Should we default the service to a value, emit a null service tag
+  or not include the tag at all if there is no service?
```

Comment:
> Should the default be to not emit tags (source and service)? Could the default be different for OP users and standard Vector users?

This has product implications so Barry can speak to those aspecta of it, but seems we can enforce it on the OP side to make sure it’s consistent for customers (and can be relied upon in the API/UI).  This would be a settings that we can switch on/off on internal metrics?  So when we setup this pipeline internal pipeline we can insure the tags are being set?
https://github.com/DataDog/observability-pipelines-worker/blob/main/src/configuration/internal_pipeline.rs#L44-L52
I would say having a setting for enabling each source and service would give us flexibility.
> What should the source be tagged as for events that go through a transform that could potentially combine one event from multiple sources - eg. reduce?

Let Barry give the product take on this.  Right.  Are there other scenarios you can think of?  For reduce I think I would take the first source id that enters the transaction or some other stable logic (based on what’s easy to implement).  It seems we could recommend not crossing source boundaries if people want better visibility for these cases.
> What should the source be for events that come from no source (emitted by the lua transform)?

Are there other examples of this?  Lua is probably not worth spending a lot of time on on.  Could we just use an internal tag value like   _no_source  or _unknown ? At least we could bucket them and handle them in some way in the API/UI.
> Not all sources will have a service. Should we default the service to a value, emit a null service tag  or not include the tag at all if there is no service?

Similar to above, my :twocents: is that we should use an internal tag value like _no_service (or _none) instead of null tags.  But maybe there’s a convention or best practice around this.  Null feels like some code forgot to set the value or something rather it being a deliberate choice.

Reply:
from @barieom 

> Should the default be to not emit tags (source and service)? Could the default be different for OP users and standard Vector users?
This has product implications so Barry can speak to those aspect of it, but seems we can enforce it on the OP side to make sure it’s consistent for customers (and can be relied upon in the API/UI).  This would be a settings that we can switch on/off on internal metrics?  So when we setup this pipeline internal pipeline we can insure the tags are being set?
https://github.com/DataDog/observability-pipelines-worker/blob/main/src/configuration/internal_pipeline.rs#L44-L52

Yeah having it turned on by default for OP users and turned off by default for standard Vector users, as it may have some performance implications make sense. Speaking of, could you share any numbers around performance regressions when this gets implemented?
Also to add more context on how this feature would evolve - users should be able to define whatever key_field they want to bucket their volume insights by in OP (likely in the bootstrap.yaml or elsewhere), rather than supporting this just for the .service field like we’re doing now/in V1
> What should the source be tagged as for events that go through a transform that could potentially combine one event from multiple sources - eg. reduce?

Like Steve is asking, are scenarios outside of reduce where this could happen?
In the case of reduce, this seems like an edge case and not something many users would do (cc/ 
[@Ari Adair](https://dd.slack.com/team/U02MM3ABX2L) [@Jesse Szwedko](https://dd.slack.com/team/U01NEAWEB4G)
 in case you’ve seen otherwise and if you can share examples if so), but defaulting to source ID with the most number of events reduced into the single event then defaulting to first source id makes sense
> What should the source be for events that come from no source (emitted by the lua transform)?
Are there other examples of this?  Lua is probably not worth spending a lot of time on on.  Could we just use an internal tag value like   _no_source  or _unknown ? At least we could bucket them and handle them in some way in the API/UI.

+1 to Steve here. _unknown or from_<transform_ID> make sense
> Not all sources will have a service. Should we default the service to a value, emit a null service tag  or not include the tag at all if there is no service?
Similar to above, my :twocents: is that we should use an internal tag value like _no_service (or _none) instead of null tags.  But maybe there’s a convention or best practice around this.  Null feels like some code forgot to set the value or something rather it being a deliberate choice.

Also +1 to Steve